src/api/models/revisions.py

```diff
@@ -0,0 +1,410 @@
+
+
+# -- TRIGGER SQL --
+
+forward_sql = """
+    CREATE OR REPLACE FUNCTION primary_key_column_name (table_name text)
+    RETURNS varchar AS $$
+    DECLARE
+    column_name text;
+    BEGIN
+        SELECT
+            pg_attribute.attname
+        INTO
+            column_name
+        FROM pg_index, pg_class, pg_attribute, pg_namespace
+        WHERE
+            pg_class.oid = table_name::regclass AND
+            indrelid = pg_class.oid AND
+            nspname = 'public' AND
+            pg_class.relnamespace = pg_namespace.oid AND
+            pg_attribute.attrelid = pg_class.oid AND
+            pg_attribute.attnum = any(pg_index.indkey)
+            AND indisprimary;
+        RETURN column_name;
+
+    END;
+    $$
+    LANGUAGE plpgsql;
+
+    CREATE OR REPLACE FUNCTION record_change()
+    RETURNS TRIGGER LANGUAGE PLPGSQL AS $$
+
+    DECLARE
+        pk uuid;
+        record_revision_id int;
+        rev_id_val uuid;
+        profile_id_val uuid;
+        project_id_val uuid;
+        is_deleted boolean;
+        updated_on_val timestamp;
+        pk_col_name varchar;
+        record jsonb;
+
+    BEGIN
+        pk_col_name := primary_key_column_name(TG_TABLE_NAME);
+        updated_on_val := now();
+        rev_id_val := uuid_generate_v4();
+
+        IF TG_OP = 'DELETE' THEN
+            record := to_jsonb(OLD);
+            is_deleted := true;
+        ELSE
+            record := to_jsonb(NEW);
+            is_deleted := false;
+        END IF;
+
+        pk := record->>pk_col_name;
+
+        IF record ? 'profile_id' THEN
+            profile_id_val := NEW.profile_id;
+        ELSE
+            profile_id_val := null;
+        END IF;
+
+        IF record ? 'project_id' THEN
+            project_id_val := NEW.project_id;
+        ELSE
+            project_id_val := null;
+        END IF;
+
+        INSERT INTO record_revision (
+            "rev_id",
+            "table_name",
+            "record_id",
+            "project_id",
+            "profile_id",
+            "updated_on",
+            "deleted"
+        )
+        VALUES (
+            rev_id_val,
+            TG_TABLE_NAME,
+            pk,
+            project_id_val,
+            profile_id_val,
+            updated_on_val,
+            is_deleted
+        )
+        ON CONFLICT (table_name, record_id) DO
+        UPDATE SET
+            "rev_id" = rev_id_val,
+            "updated_on" = updated_on_val,
+            "project_id" = project_id_val,
+            "profile_id" = profile_id_val,
+            "deleted" = is_deleted;
+
+        SELECT
+            id 
+        INTO
+            record_revision_id
+        FROM 
+            record_revision
+        WHERE
+            rev_id = rev_id_val;
+
+        /*
+          Update table revsions
+        */
+
+        INSERT INTO table_revision(
+            "last_revision_id"
+        )
+        VALUES (
+            record_revision_id
```

Comment:
Do we need `record_revision_id` at all? Can't we just use `rev_id_val` here, and remove the previous `SELECT`?